File: src/domain/reporting.ts
import { addMonths, daysBetween, monthKey, startOfMonth } from './dates';
import { checkManningCompliance, getOverdueDays } from './rules';
import { type Assignment, type Certification, type CrewMember, type IsoDate, type Rank, RANKS, type Vessel } from './types';

export const EXPIRING_SOON_DAYS = 30;
const TREND_MONTHS = 12;

export type ExpiryBucket = 'Expired' | 'Within 30 days' | 'Within 90 days' | 'Valid';

/** The colour-coding the certification views key off. */
export function expiryBucket(expiryDate: IsoDate, today: IsoDate): ExpiryBucket {
  const days = daysBetween(today, expiryDate);
  if (days < 0) return 'Expired';
  if (days <= 30) return 'Within 30 days';
  if (days <= 90) return 'Within 90 days';
  return 'Valid';
}

export interface DashboardSummary {
  crewOnboard: number;
  vesselsBelowManning: number;
  certificationsExpiringSoon: number;
  overdueRotations: number;
  crewByRank: { rank: Rank; count: number }[];
  rotationsOverTime: { month: string; signOns: number; signOffs: number }[];
  fleetCompliance: { compliant: number; belowManning: number };
  certificationStatus: { bucket: ExpiryBucket; count: number }[];
}

// Every dashboard figure, computed in one pass over the data. Pure, so it's
// testable without a database. Reuses the business rules rather than
// reimplementing "below manning" or "overdue", so the dashboard can't drift
// from the pages it links to.
export function buildDashboardSummary(
  vessels: readonly Vessel[],
  crew: readonly CrewMember[],
  assignments: readonly Assignment[],
  certifications: readonly Certification[],
  today: IsoDate,
): DashboardSummary {
  const activeAssignments = assignments.filter(
    (assignment) =>
      assignment.status === 'Active' &&
      assignment.signOnDate <= today &&
      today <= assignment.signOffDate,
  );

  const belowManning = vessels.filter(
    (vessel) => !checkManningCompliance(vessel, assignments, today).compliant,
  ).length;

  const rankCounts = new Map<Rank, number>();
  for (const member of crew) {
    rankCounts.set(member.rank, (rankCounts.get(member.rank) ?? 0) + 1);
  }

  const bucketCounts = new Map<ExpiryBucket, number>();
  let expiringSoon = 0;
  for (const certification of certifications) {
    const bucket = expiryBucket(certification.expiryDate, today);
    bucketCounts.set(bucket, (bucketCounts.get(bucket) ?? 0) + 1);
    const days = daysBetween(today, certification.expiryDate);
    if (days >= 0 && days <= EXPIRING_SOON_DAYS) expiringSoon += 1;
  }

  // A fixed window of months, so a month with no movements still appears as a
  // zero rather than vanishing and distorting the shape of the line.
  const firstMonth = startOfMonth(addMonths(today, -(TREND_MONTHS - 1)));
  const months = Array.from({ length: TREND_MONTHS }, (_unused, index) =>
    monthKey(addMonths(firstMonth, index)),
  );
  const signOns = new Map<string, number>();
  const signOffs = new Map<string, number>();
  for (const assignment of assignments) {
    const onKey = monthKey(assignment.signOnDate);
    const offKey = monthKey(assignment.signOffDate);
    signOns.set(onKey, (signOns.get(onKey) ?? 0) + 1);
    signOffs.set(offKey, (signOffs.get(offKey) ?? 0) + 1);
  }

  return {
    crewOnboard: new Set(activeAssignments.map((assignment) => assignment.crewId)).size,
    vesselsBelowManning: belowManning,
    certificationsExpiringSoon: expiringSoon,
    overdueRotations: assignments.filter((assignment) => getOverdueDays(assignment, today) > 0)
      .length,
    crewByRank: RANKS.map((rank) => ({ rank, count: rankCounts.get(rank) ?? 0 })),
    rotationsOverTime: months.map((month) => ({
      month,
      signOns: signOns.get(month) ?? 0,
      signOffs: signOffs.get(month) ?? 0,
    })),
    fleetCompliance: { compliant: vessels.length - belowManning, belowManning },
    certificationStatus: (
      ['Expired', 'Within 30 days', 'Within 90 days', 'Valid'] as const
    ).map((bucket) => ({ bucket, count: bucketCounts.get(bucket) ?? 0 })),
  };
}

File: server/CrewLink.Api/Domain/Reporting.cs
using System.Text.Json.Serialization;

namespace CrewLink.Api.Domain;

public enum ExpiryBucket
{
    [JsonStringEnumMemberName("Expired")] Expired,
    [JsonStringEnumMemberName("Within 30 days")] Within30Days,
    [JsonStringEnumMemberName("Within 90 days")] Within90Days,
    [JsonStringEnumMemberName("Valid")] Valid,
}

public sealed record RankCount(Rank Rank, int Count);

public sealed record RotationsInMonth(string Month, int SignOns, int SignOffs);

public sealed record FleetCompliance(int Compliant, int BelowManning);

public sealed record CertificationStatusCount(ExpiryBucket Bucket, int Count);

public sealed record DashboardSummary(
    int CrewOnboard,
    int VesselsBelowManning,
    int CertificationsExpiringSoon,
    int OverdueRotations,
    IReadOnlyList<RankCount> CrewByRank,
    IReadOnlyList<RotationsInMonth> RotationsOverTime,
    FleetCompliance FleetCompliance,
    IReadOnlyList<CertificationStatusCount> CertificationStatus);

public static class Reporting
{
    public const int ExpiringSoonDays = 30;
    private const int TrendMonths = 12;

    /// <summary>The colour-coding the certification views key off.</summary>
    public static ExpiryBucket GetExpiryBucket(DateOnly expiryDate, DateOnly today)
    {
        var days = Dates.DaysBetween(today, expiryDate);
        if (days < 0) return ExpiryBucket.Expired;
        if (days <= 30) return ExpiryBucket.Within30Days;
        if (days <= 90) return ExpiryBucket.Within90Days;
        return ExpiryBucket.Valid;
    }

    /// <summary>Every dashboard figure, computed in one pass over the data.</summary>
    /// <remarks>
    /// Reuses the business rules rather than reimplementing "below manning" or
    /// "overdue" — two definitions of the same concept is how a dashboard ends
    /// up disagreeing with the page it links to.
    /// </remarks>
    public static DashboardSummary BuildDashboardSummary(
        IReadOnlyList<Vessel> vessels,
        IReadOnlyList<CrewMember> crew,
        IReadOnlyList<Assignment> assignments,
        IReadOnlyList<Certification> certifications,
        DateOnly today)
    {
        // Distinct crew, not active assignments — a data error producing two
        // active rows for one person shouldn't count as two.
        var crewOnboard = assignments
            .Where(assignment =>
                assignment.Status == AssignmentStatus.Active &&
                assignment.SignOnDate <= today &&
                today <= assignment.SignOffDate)
            .Select(assignment => assignment.CrewId)
            .Distinct()
            .Count();

        var belowManning = vessels.Count(vessel =>
            !Rules.CheckManningCompliance(vessel, assignments, today).Compliant);

        var rankCounts = new Dictionary<Rank, int>();
        foreach (var member in crew)
        {
            rankCounts[member.Rank] = rankCounts.GetValueOrDefault(member.Rank) + 1;
        }

        var bucketCounts = new Dictionary<ExpiryBucket, int>();
        var expiringSoon = 0;
        foreach (var certification in certifications)
        {
            var bucket = GetExpiryBucket(certification.ExpiryDate, today);
            bucketCounts[bucket] = bucketCounts.GetValueOrDefault(bucket) + 1;

            var days = Dates.DaysBetween(today, certification.ExpiryDate);
            if (days >= 0 && days <= ExpiringSoonDays) expiringSoon += 1;
        }

        // A fixed window of months, built first and filled after, so a month
        // with no movements is a zero rather than a gap in the line.
        var firstMonth = Dates.StartOfMonth(today.AddMonths(-(TrendMonths - 1)));
        var months = Enumerable
            .Range(0, TrendMonths)
            .Select(offset => Dates.MonthKey(firstMonth.AddMonths(offset)))
            .ToList();

        var signOns = new Dictionary<string, int>(StringComparer.Ordinal);
        var signOffs = new Dictionary<string, int>(StringComparer.Ordinal);
        foreach (var assignment in assignments)
        {
            var onKey = Dates.MonthKey(assignment.SignOnDate);
            var offKey = Dates.MonthKey(assignment.SignOffDate);
            signOns[onKey] = signOns.GetValueOrDefault(onKey) + 1;
            signOffs[offKey] = signOffs.GetValueOrDefault(offKey) + 1;
        }

        return new DashboardSummary(
            CrewOnboard: crewOnboard,
            VesselsBelowManning: belowManning,
            CertificationsExpiringSoon: expiringSoon,
            OverdueRotations: assignments.Count(assignment => Rules.GetOverdueDays(assignment, today) > 0),
            CrewByRank: Ranks.All
                .Select(rank => new RankCount(rank, rankCounts.GetValueOrDefault(rank)))
                .ToList(),
            RotationsOverTime: months
                .Select(month => new RotationsInMonth(
                    month,
                    signOns.GetValueOrDefault(month),
                    signOffs.GetValueOrDefault(month)))
                .ToList(),
            FleetCompliance: new FleetCompliance(vessels.Count - belowManning, belowManning),
            CertificationStatus: EnumNames<ExpiryBucket>.Values
                .Select(bucket => new CertificationStatusCount(bucket, bucketCounts.GetValueOrDefault(bucket)))
                .ToList());
    }
}
